Add hosted version of the batch processor controller and fix related unit tests

xristos3490 opened the pull request:
## Description

This pull request brings a hosted version of core's [BatchProcessingController](https://github.com/woocommerce/woocommerce/blob/37bb36994b08d417e34b9043fea0bf92a3f9c236/plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php#L29) as `WCS_Batch_Processing_Controller` together with a new `WCS_Batch_Processor` interface. It also addressed required changes for the unit tests, adds more coverage, and fixes the provisioning for new installations.

### Reason for moving these interfaces

Before [this](https://github.com/woocommerce/woocommerce/pull/46975) fix was introduced (which was implemented in WC v9.0), the batch processing controller in Core could only work with "namespaced" processors. To avoid using polyfill classes, we decided to transfer the functionality to WCS for better stability. Plus, using internal-namespaced classes is considered an antipattern, so moving the functionality was the most direct solution.

## How to test this PR

Unit Tests 
+
Exploratory testing. Some suggested steps:
- Change the offset or enable it if it's disabled.
- Keep an eye on AS actions in the "Tools > Scheduled Actions" table.
- Observe any changes made to AS actions.
- Now, delete any notification-related AS actions.
- Go to "WooCommerce > Status > Tools" and click on the "Add notifications" tool.
- Keep an eye on the AS actions.
- Confirm that the expected AS actions are restored after the batcher has run.

## Product impact
<!-- What products will this PR ship in? -->

- [ ] Added changelog entry (or does not apply)
- [ ] Will this PR affect WooCommerce Subscriptions? yes/no/tbc, add issue ref
- [ ] Will this PR affect WooCommerce Payments? yes/no/tbc, add issue ref
- [ ] <!-- 🚨 Deprecations 🚨 --> Added deprecated functions, hooks or classes to the [spreadsheet](https://docs.google.com/spreadsheets/d/1xw9xszcPMnWsp4C8OKZMsLzZob7tOmWT7qMqmEIq314/edit#gid=0)

xristos3490 committed:
Fix debug tool tests

xristos3490 committed:
Test batch processor

xristos3490 committed:
Test the emails controller

xristos3490 committed:
Fix update time issue on first load

xristos3490 committed:
Some polishing

xristos3490 committed:
Merge branch 'try/subs-notifications-pf' into improve/batch-unit-tests

xristos3490 committed:
Fix emails controller tests

xristos3490 committed:
Fix tests and initial state of DB

xristos3490 committed:
More Cleanup;

xristos3490 committed:
Merge branch 'try/subs-notifications-pf' into improve/batch-unit-tests

xristos3490 committed:
Bring hosted version of the batch controller and interface

xristos3490 committed:
Fix tests to use the new interfaces

xristos3490 committed:
Fix the autoloader

xristos3490 committed:
Update the processors to use the new interfaces

xristos3490 committed:
Fix the main file and enable feature for new installations fix

xristos3490 committed:
Merge branch 'try/subs-notifications-pf' into improve/batch-unit-tests

xristos3490 committed:
Cleanup

xristos3490 committed:
Cleanup

xristos3490 committed:
Revert update-time rename

xristos3490 commented on includes/class-wc-subscriptions-email-notifications.php:
```diff
@@ -20,23 +20,30 @@ class WC_Subscriptions_Email_Notifications {
 
 		add_action( 'update_option_' . WC_Subscriptions_Admin::$option_prefix . self::$offset_setting_string, [ 'WC_Subscriptions_Email_Notifications', 'set_notification_settings_update_time' ] );
 		add_action( 'update_option_' . WC_Subscriptions_Admin::$option_prefix . self::$switch_setting_string, [ 'WC_Subscriptions_Email_Notifications', 'set_notification_settings_update_time' ] );
+		add_action( 'add_option_' . WC_Subscriptions_Admin::$option_prefix . self::$offset_setting_string, [ 'WC_Subscriptions_Email_Notifications', 'set_notification_settings_update_time' ] );
```
This addresses the issue where the callback is missed on the first write for existing installations.

peterfabian replied:
Well spotted!

xristos3490 commented on includes/class-wcs-batch-processing-controller.php:
```diff
@@ -0,0 +1,572 @@
+ * @category Class
+ * @since    8.0.0
+ */
+class WCS_Batch_Processing_Controller {
```
I tried to simplify this, but it seems pretty abstracted, already!

peterfabian commented on includes/admin/debug-tools/class-wcs-notifications-debug-tool-processor.php:
```diff
@@ -210,7 +214,7 @@ public function get_next_batch_to_process( int $size ): array {
 	public function process_batch( array $batch ): void {
-		// This is a bit unnecessary. Perhaps convert `update_status` to static to avoid instantiating the class?
+
 		$subscriptions_notifications = new WCS_Action_Scheduler_Customer_Notifications();
```
Just a note here that I removed the instantiation in WCS_Notifications_Batch_Processor because `WCS_Action_Scheduler_Customer_Notifications` hooks `update_status`, `update_date` and `delete_date` via its parent `WCS_Scheduler` and wanted to avoid having double/triple hooks executed (even though they were implemented to be idempotent). Please see https://github.com/Automattic/woocommerce-subscriptions-core/pull/648/files#diff-d9eda2054181e7a8de782ad735d85bb637f35ad5c67dee9262b6518578cf6eb0R172

I think we should do the same here (not necessarily in this PR, but let's not forget)

xristos3490 replied:
Done in https://github.com/Automattic/woocommerce-subscriptions-core/pull/689/commits/2332026e77506e4a944ae51dc3d92102704e6f18

peterfabian commented on includes/admin/debug-tools/class-wcs-notifications-debug-tool-processor.php:
```diff
@@ -210,7 +214,7 @@ public function get_next_batch_to_process( int $size ): array {
-		// This is a bit unnecessary. Perhaps convert `update_status` to static to avoid instantiating the class?
+
 		$subscriptions_notifications = new WCS_Action_Scheduler_Customer_Notifications();
 
```
Also, I suppose to be aligned with the approach taken in the ad-hoc batch processor, we should unschedule all notifications if they're disabled? 
https://github.com/Automattic/woocommerce-subscriptions-core/pull/648/files#diff-d9eda2054181e7a8de782ad735d85bb637f35ad5c67dee9262b6518578cf6eb0R176-R180

xristos3490 replied:
Done https://github.com/Automattic/woocommerce-subscriptions-core/pull/689/commits/2332026e77506e4a944ae51dc3d92102704e6f18

xristos3490 replied:
@peterfabian I might be missing something, but could we avoid using a processor altogether and instead just remove every AS action for a given group via `as_unschedule_all_actions()`. Essentially, calling the `ActionScheduler_Store::instance()->cancel_actions_by_group()` behind the scenes? 

Same for the batch processor obviously.

peterfabian replied:
No, I don't think you're missing anything. You're right that fewer abstractions and function calls are better. We can just call `as_unschedule_all_actions( WCS_Action_Scheduler_Customer_Notifications::$notifications_as_group )`. 

Essentially I just used it to encapsulate the group so that it's not a public property, but we can probably just make it public and call the AS function directly.

peterfabian commented on includes/admin/debug-tools/class-wcs-notifications-debug-tool-processor.php:
```diff
@@ -301,7 +308,7 @@ public function handle_woocommerce_debug_tools( array $tools ): array {
 				'desc'     =>
 				/* translators: %d=count of entries pending conversion */
-					sprintf( __( 'This tool will add notifications to pending, active and on-hold subscriptions. This will happen overtime in the background (via Action Scheduler). There are currently %d entries that can be converted.', 'woocommerce-subscriptions' ), $pending_count ),
+					sprintf( __( 'This tool will add notifications to pending, active, and on-hold subscriptions. These updates will occur gradually in the background using Action Scheduler. Currently, there are %d entries available for conversion.', 'woocommerce-subscriptions' ), $pending_count ),
```
I don't think we should call it conversion. Maybe sth like "Currently, there are %d subscriptions to process."

xristos3490 replied:
Thanks! Done in https://github.com/Automattic/woocommerce-subscriptions-core/pull/689/commits/b564854949cfbe6694287c9c3e7f9a835b2591f0

We might want to revisit during polishing!

peterfabian commented on includes/class-wc-subscriptions-core-plugin.php:
```diff
@@ -164,10 +164,14 @@ public function init() {
 		$scheduler_class = apply_filters( 'woocommerce_subscriptions_scheduler', 'WCS_Action_Scheduler' );
 		$this->scheduler = new $scheduler_class();
 
+		// Custom notifications scheduler.
```
Custom or Customer? :) Or Custom customer notifications scheduler? :)

peterfabian commented on includes/class-wc-subscriptions-core-plugin.php:
```diff
@@ -529,16 +533,9 @@ public function activate_plugin() {
 
-			// If this is the first time activating WooCommerce Subscription we want to enable the customer email notifications (default to 3 days before.)
-			if ( '0' === get_option( WC_Subscriptions_Admin::$option_prefix . '_previous_version', '0' ) && false === get_option( WC_Subscriptions_Admin::$option_prefix . WC_Subscriptions_Email_Notifications::$offset_setting_string, false ) ) {
+			// Enable customer notifications by default.
```
```suggestion
			// Enable customer notifications by default for new stores.
```
This is only for new stores, right?

peterfabian commented on includes/class-wc-subscriptions-email-notifications.php:
```diff
@@ -189,16 +208,17 @@ public static function should_send_notification() {
 
 		// By default, there are no shorter periods than days in WCS, so we ignore hours, minutes, etc.
-		if ( $period <= 2 && 'day' === $interval ) {
+		if ( $interval <= 2 && 'day' === $period ) {
```
Nice catch! 🙌

peterfabian commented on includes/class-wcs-batch-processing-controller.php:
```diff
@@ -0,0 +1,572 @@
+
+	const ENQUEUED_PROCESSORS_OPTION_NAME = 'wcs_pending_batch_processes';
+	const ACTION_GROUP                    = 'wcs_batch_processes';
+	const LOGS_CONTEXT                    = 'wcs-batch-processing';
```
Nice touch!

peterfabian commented on includes/class-wcs-notifications-batch-processor.php:
```diff
@@ -33,9 +42,18 @@ protected function get_subscription_statuses() {
+			return '';
+		}
+
+		$notification_settings_update_time = new DateTime( "@$notification_settings_update_timestamp", new DateTimeZone( 'UTC' ) );
```
Thanks for fixing this!

peterfabian commented on includes/interfaces/interface-wcs-batch-processor.php:
```diff
@@ -0,0 +1,81 @@
+ *
+ * @package  WooCommerce Subscriptions
+ * @version  8.0.0
+ * @since    8.0.0
```
Side note: do we already know the version? Perhaps we should add it to more of the new code..

xristos3490 replied:
Not sure tbh. Reverted to x.x.x in https://github.com/Automattic/woocommerce-subscriptions-core/pull/689/commits/38dd921e19b8f47452dea5d8e7d34bcd9d0f1ab5

peterfabian commented on tests/unit/test-wcs-notifications-debug-tool.php:
```diff
@@ -1,9 +1,34 @@
+		$process_next_batch_for_single_processor->setAccessible( true );
+		$process_next_batch_for_single_processor->invoke( $batch_processor, WCS_Notifications_Debug_Tool_Processor::class );
+
+		$this->assertFalse( $batch_processor->is_enqueued( WCS_Notifications_Debug_Tool_Processor::class ) );
```
I suppose this means the processing has finished and the BatchProcessor dequeued itself, right? Might be good to add a comment here on why we assume this.

xristos3490 replied:
Good idea! Done https://github.com/Automattic/woocommerce-subscriptions-core/pull/689/commits/e9e46076399926af6d0188cb41ccd8b3577b785f

peterfabian commented on tests/unit/test-wcs-notifications-debug-tool.php:
```diff
@@ -1,9 +1,34 @@
+	 * @param array $data
+	 * @return bool
+	 */
+	public function test_batch_processesing_controller() {
```
```suggestion
	public function test_batch_processing_controller() {
```

peterfabian commented on tests/unit/test-wcs-notifications-debug-tool.php:
```diff
@@ -1,9 +1,34 @@
+	 *
+	 * @covers WCS_Notifications_Debug_Tool_Processor::process_batch
+	 *
+	 * @param array $data
```
```suggestion

```

peterfabian commented on tests/unit/test-wcs-notifications-debug-tool.php:
```diff
@@ -1,9 +1,34 @@
+	 * @covers WCS_Notifications_Debug_Tool_Processor::process_batch
+	 *
+	 * @param array $data
+	 * @return bool
```
```suggestion
	 * @return void
```

peterfabian commented on tests/unit/test-wcs-notifications-debug-tool.php:
```diff
@@ -212,4 +155,78 @@ public function test_tool_state_while_processing() {
+
+		$simple_subscription->update_dates(
+			[
+				'next_payment' => '2034-09-20 08:08:08',
```
Is this supposed to be "always" in the future? Why not use a relative date, then?

peterfabian replied:
Also, is it a problem that 1-month subscription has next payment in 10 year's time? :)

xristos3490 replied:
> Is this supposed to be "always" in the future? Why not use a relative date, then?

Yes, I suppose I can make these dates dynamic. I wouldn't want someone 10 years from now cursing me while digging into a nasty unit test bug. 😅 

> Also, is it a problem that 1-month subscription has next payment in 10 year's time? :)

Yes, it seems odd. I believe it's related to our recent conversation about the differences between a "checkout" flow and the freestyle data model, which I actually don't mind.

xristos3490 replied:
Done https://github.com/Automattic/woocommerce-subscriptions-core/pull/689/commits/f9bee572a8cdd2f32271bfa2a4a05b743ca7ba54 and https://github.com/Automattic/woocommerce-subscriptions-core/pull/689/commits/725f13334e230eec6cfc0352dd123c5b111ba232

peterfabian commented on tests/unit/test-wcs-notifications-debug-tool.php:
```diff
@@ -14,17 +39,8 @@ class WCS_Subscription_Notifications_Debug_Tool_Test extends WP_UnitTestCase {
-
-		$batches = $this->notification_subscription_provider();
+		$batches   = $this->notification_subscription_data_provider();
+		$processor = new WCS_Notifications_Debug_Tool_Processor();
```
Might be better to call `\WCS_Action_Scheduler_Customer_Notifications::get_action_args` in the `$action_args` below.

xristos3490 replied:
Done https://github.com/Automattic/woocommerce-subscriptions-core/pull/689/commits/f9bee572a8cdd2f32271bfa2a4a05b743ca7ba54

peterfabian commented on tests/unit/test-wcs-notifications-emails.php:
```diff
@@ -0,0 +1,118 @@
+	}
+
+	/**
+	 * Test subscription period too long.
```
```suggestion
	 * Test global subscription notification switch.
```

xristos3490 replied:
Haha! AI FTW! Nice catch; funny one!

xristos3490 replied:
Done https://github.com/Automattic/woocommerce-subscriptions-core/pull/689/commits/f9bee572a8cdd2f32271bfa2a4a05b743ca7ba54

peterfabian commented on tests/unit/test-wcs-notifications-emails.php:
```diff
@@ -0,0 +1,118 @@
+
+	/**
+	 * Helper to enable notifications globally.
+	 * TODO: We should create global helpers?
```
Yes :)

xristos3490 replied:
Perhaps we can do that alongside your tests! I'll follow up on a PR after this gets merged?

peterfabian replied:
Sounds good.

peterfabian reviewed (commented):
Overall, this looks nice! 

How do you see the robustness of the batch processing?

I'm thinking about scenario when one of the subscriptions will fail to get processed because it's in some invalid state or something (and an exception is thrown in the foreach loop). 

Then `process_batch` could theoretically never update the `last_offset` and total pending count will never go down, leaving the debug tool spinning forever.

I suppose we might have the same problem with the `WCS_Notifications_Batch_Processor` class if it wouldn't manage to update the subscription's update time...

xristos3490 committed:
Use the singleton and update logic to unschedule entities

xristos3490 committed:
Improve debug tool copy

xristos3490 committed:
Improve tests and docblocks

xristos3490 committed:
More relative dates to the batch processor tests

xristos3490 committed:
Revert v8 to x.x.x

xristos3490 committed:
More test comments

xristos3490 commented:
Thank you for the review, @peterfabian! I have addressed all of your feedback. Could you please take another look? 🙌 

> I'm thinking about scenario when one of the subscriptions will fail to get processed because it's in some invalid state or something (and an exception is thrown in the foreach loop).

Assuming that the `WCS_Batch_Processing_Controller` error handling catches the errors, each processor should eventually pass the `is_consistently_failing()` check and dequeue itself with a log message, right?

Given that there is no condition to be met for our "flags" to be modified and thus prevent an infinite loop, the only reason would be a fatal error, resulting in the system being in a constantly-failing mode.

peterfabian reviewed (approved):
All the points I raised have been addressed, so let's merge this! 
Great work 💪 !

Extra kudos for making the tests pass 👌
